graph/src/loader.c

```diff
@@ -137,25 +137,38 @@ void rage_element_loader_unload(
     }
 }
 
+// ConcreteElementType
+
+rage_NewConcreteElementType rage_element_type_specialise(
+        rage_ElementType * type, rage_Atom ** params) {
+    rage_NewInstanceSpec new_ports = type->get_ports(params);
+    RAGE_EXTRACT_VALUE(
+	rage_NewConcreteElementType, new_ports, rage_InstanceSpec spec)
+    rage_ConcreteElementType * cet = malloc(sizeof(rage_ConcreteElementType));
+    cet->type = type;
+    cet->params = params;
+    cet->spec = spec;
+    return RAGE_SUCCESS(rage_NewConcreteElementType, cet);
+}
+
+void rage_concrete_element_type_free(rage_ConcreteElementType * cet) {
+    cet->type->free_ports(cet->spec);
+    free(cet);
+}
+
+// Element
+
 rage_ElementNewResult rage_element_new(
-        rage_ElementType * type, uint32_t sample_rate, uint32_t frame_size,
-        rage_Atom ** params) {
-    rage_NewElementState new_state = type->state_new(
-        sample_rate, frame_size, params);
+        rage_ConcreteElementType * cet, uint32_t sample_rate,
+        uint32_t frame_size) {
+    rage_NewElementState new_state = cet->type->state_new(
+        sample_rate, frame_size, cet->params);
     RAGE_EXTRACT_VALUE(rage_ElementNewResult, new_state, void * state)
     rage_Element * const elem = malloc(sizeof(rage_Element));
-    elem->type = type;
+    elem->type = cet->type;
     elem->state = state;
-    rage_NewInstanceSpec new_ports = type->get_ports(params);
-    if (RAGE_FAILED(new_ports)) {
```

Comment:
Have we lost the error checking in moving this around? Or am I losing myself in the macros?

Reply:
The EXTRACT_VALUE macro exits on failure.
It isn't very clear or generally applicable though, and as it has got simpler due to the introduction of helpers for parts of what it does I have been wondering about ditching it.